azure_provisioning_e2e/tests/test_sync_certificate_enrollments.py

```diff
@@ -0,0 +1,266 @@
+            device_index=device_index, device_id=device_id
+        )
+        registration_id = individual_enrollment_record.registration_id
+
+        device_cert_file = "demoCA/newcerts/device_cert" + str(device_index) + ".pem"
+        device_key_file = "demoCA/private/device_key" + str(device_index) + ".pem"
+        result_from_register(registration_id, device_cert_file, device_key_file)
+
+        assert_device_provisioned(device_id)
+    finally:
+        service_client.delete_individual_enrollment_by_param(registration_id)
+
+
+@pytest.mark.it(
+    "A device gets provisioned to the linked IoTHub with device_id equal to the registration_id of the individual enrollment that has been created with a selfsigned X509 authentication"
+)
+def test_device_register_with_no_device_id_for_a_x509_individual_enrollment(before_all_tests):
+
+    device_index = 2
+
+    try:
+        individual_enrollment_record = create_individual_enrollment_with_x509_client_certs(
+            device_index=device_index
+        )
+        registration_id = individual_enrollment_record.registration_id
+
+        device_cert_file = "demoCA/newcerts/device_cert" + str(device_index) + ".pem"
+        device_key_file = "demoCA/private/device_key" + str(device_index) + ".pem"
+        result_from_register(registration_id, device_cert_file, device_key_file)
+
+        assert_device_provisioned(registration_id)
+    finally:
+        service_client.delete_individual_enrollment_by_param(registration_id)
+
+
+@pytest.mark.it(
+    "A group of devices get provisioned to the linked IoTHub with device_ids equal to the individual registration_ids inside a group enrollment that has been created with intermediate X509 authentication"
+)
+def test_group_of_devices_register_with_no_device_id_for_a_x509_intermediate_authentication_group_enrollment(
+    before_all_tests
+):
+    group_id = "e2e-intermediate-hogwarts"
+    common_device_id = device_common_name
+    device_count_in_group = 3
+    reprovision_policy = ReprovisionPolicy(migrate_device_data=True)
+
+    try:
+        intermediate_cert_filename = "demoCA/newcerts/intermediate_cert.pem"
+        with open(intermediate_cert_filename, "r") as intermediate_pem:
+            intermediate_cert_content = intermediate_pem.read()
+
+        attestation_mechanism = AttestationMechanism.create_with_x509_signing_certs(
+            intermediate_cert_content
+        )
+        enrollment_group_provisioning_model = EnrollmentGroup.create(
+            group_id, attestation=attestation_mechanism, reprovision_policy=reprovision_policy
+        )
+
+        service_client.create_or_update(enrollment_group_provisioning_model)
+
+        count = 0
+        common_device_key_input_file = "demoCA/private/device_key"
+        common_device_cert_input_file = "demoCA/newcerts/device_cert"
+        common_device_inter_cert_chain_file = "demoCA/newcerts/out_inter_device_chain_cert"
+        for index in range(3, 3 + device_count_in_group):
```

Comment:
>3 [](start = 27, length = 1)

This is because we've already used device index 1 and 2 right?

Manually maintaining device indicies for each test in a constant is probably not a great idea since it can go wrong in a lot of places. Could you perhaps track them in some kind of global dict/map?